discover-providers.js

```diff
@@ -73,13 +77,63 @@ function createJsonFileDiscoverProvider(hostsFile) {
 }
 
 /**
+ * Wrap a innerDiscoverProvider with a retrying mechanism.
+ *
+ * @param {Object} opts the options used as a Backoff policy {@see Backoff}
+ * @param {Object} [opts.backoff] a mocked backoff for testing
+ * @param {DiscoverProvider} innerDiscoverProvider The discover provider to wrap
+ *
+ * @returns {DiscoverProvider} The retrying discover provider.
+ */
+function retryDiscoverProvider(opts, innerDiscoverProvider) {
+    if (!opts) {
+        return innerDiscoverProvider;
+    }
+
+    return function retryingDiscoverProvider(callback) {
+        var policy;
+        var defaultPolicy = {
+            minDelay: 500,
+            maxDelay: 15000,
+            timeout: 60000
+        };
+
+        if (opts === true) {
+            policy = defaultPolicy;
+        } else {
+            policy = _.defaults({}, opts, defaultPolicy);
+        }
+        var backoff = opts.backoff || new Backoff(policy);
+        var lastError = null;
+
+        attempt(null);
+
+        function attempt(fail) {
+            if (fail) {
+                return setImmediate(callback, lastError);
```

Comment:
`lastError` might be `null` if the last call to the `innerDiscoveryProvider` didn't result in an error but in an empty host list. The result is that the `callback` that is passed to the discovery provider will return without a hosts list (`undefined` compared to an empty list) and without an error. Are we sure that the consumer of a call to the discovery provider can handle these cases?

I think it is safe to pass an empty list always to the callback to make sure it will not try to iterate over an undefined list.